source/extensions/filters/http/thrift_to_metadata/filter.cc

```diff
@@ -10,11 +10,356 @@ namespace Extensions {
+      respstats_{ALL_THRIFT_TO_METADATA_FILTER_STATS(
+          POOL_COUNTER_PREFIX(scope, "thrift_to_metadata.resp"))},
+      request_rules_(generateRules(proto_config.request_rules())),
+      response_rules_(generateRules(proto_config.response_rules())),
+      transport_(ProtoUtils::getTransportType(proto_config.transport())),
+      protocol_(ProtoUtils::getProtocolType(proto_config.protocol())),
+      allow_content_types_(generateAllowContentTypes(proto_config.allow_content_types())),
+      allow_empty_content_type_(proto_config.allow_empty_content_type()) {
+  if (request_rules_.empty() && response_rules_.empty()) {
+    throw EnvoyException("thrift_to_metadata filter: Per filter configs must at least specify "
+                         "either request or response rules");
+  }
+
+  if (proto_config.protocol() == envoy::extensions::filters::network::thrift_proxy::v3::TWITTER) {
+    throw EnvoyException("thrift_to_metadata filter: Protocol TWITTER is not supported");
+  }
+}
+
+bool FilterConfig::contentTypeAllowed(absl::string_view content_type) const {
+  if (content_type.empty()) {
+    return allow_empty_content_type_;
+  }
+
+  return allow_content_types_.contains(content_type);
+}
+
+Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool end_stream) {
+  if (!config_->doRequestMetadata()) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  if (!config_->contentTypeAllowed(headers.getContentTypeValue())) {
+    request_processing_finished_ = true;
+    config_->rqstats().mismatched_content_type_.inc();
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  if (end_stream) {
+    handleAllOnMissing(config_->requestRules(), *decoder_callbacks_, request_processing_finished_);
+    config_->rqstats().no_body_.inc();
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  rq_handler_ = config_->createThriftDecoderHandler(*this, true);
+  return Http::FilterHeadersStatus::StopIteration;
+}
+
+Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_stream) {
+  if (!config_->doRequestMetadata() || request_processing_finished_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  // Set ``request_processing_finished_`` once we get enough data to trigger messageBegin.
+  if (!processData(data, rq_buffer_, rq_handler_, *decoder_callbacks_)) {
+    handleAllOnMissing(config_->requestRules(), *decoder_callbacks_, request_processing_finished_);
+    config_->rqstats().invalid_thrift_body_.inc();
+    return Http::FilterDataStatus::Continue;
+  }
+
+  // Handle incomplete request thrift message.
+  if (end_stream && !request_processing_finished_) {
+    ENVOY_LOG(trace,
+              "thrift to metadata filter decodeData: handle incomplete request thrift message");
+
+    handleAllOnMissing(config_->requestRules(), *decoder_callbacks_, request_processing_finished_);
+    config_->rqstats().invalid_thrift_body_.inc();
+  }
```

Comment:
I think in this case you should return `Http::FilterDataStatus::Continue` because you should never see another call with more data and since `request_processing_finished_` is false the function will return StopIterationAndBuffer.

Reply:
actually `handleAllOnMissing` will call `finalizeDynamicMetadata` which sets the flag `request_processing_finished_`

but i guess it's more readable to explicit return `Continue`